### PR TITLE
Unify args validation for discrete and continuous distributions

### PR DESCRIPTION
**Pull request by neerajprad**
This refactor addresses the following:
 - Add some more constraints for discrete distributions.
 - Removes existing `_args_check` logic to instead use the `_validate_args` checks from @fehiepsi's earlier PRs #98. 
 - Changes the default behavior to not validate (this is something that can be changed later). Changes the `validation_disabled` contextmanager to `validation_enabled` and moves this to `distribution.py` so as to avoid circular dependencies.

Right now, some dependent constraints cannot be checked, e.g. for `logits` vs `probs`. That will require these properties to be set at the instance level, and it gets complicated with the existing distributions hierarchy. Also note that this doesn't handle the multinomial distribution which is complicated for similar reasons. All of these should be addressed once #82 is fixed. 

Fixes #50.

### CODE REVIEW

File: numpyro/distributions/discrete.py [hunk 1]
```diff
@@ -33,6 +37,9 @@ def _entropy(self, n, p):
 
 
 class _bernoulli_gen(jax_discrete, bernoulli_gen):
+    # TODO: This should depend on initialization
+    arg_constraints = {'p': constraints.real}
```
Comment (fehiepsi) [inline numpyro/distributions/discrete.py]: To me, the simplest solution right now is to create a new class, named for example: bernoulli_with_logits. ^^! WDYT?

File: numpyro/distributions/distribution.py [hunk 2]
```diff
@@ -127,12 +128,13 @@ def logpdf(self, x, *args, **kwargs):
...
 
-class jax_discrete(osp_stats.rv_discrete):
-    args_check = True
-
+class jax_discrete(jax_generic, osp_stats.rv_discrete):
     def _support_mask(self, k):
         return (k >= self.a) & (k <= self.b) & (np.floor(k) == k)
 
+    def _support(self, *args, **kwargs):
+        return self._support_mask
```
Comment (fehiepsi) [inline numpyro/distributions/distribution.py]: Though it is not important because we don't want to work with `loc!=0`, could you add a TODO comment here to remind us in case this scipy interface works well on its own?

File: numpyro/distributions/constraints.py [hunk 3]
```diff
@@ -67,9 +83,12 @@ def __call__(self, x):
         return np.all(x > 0, axis=-1) & (x_sum <= 1) & (x_sum > 1 - 1e-6)
 
 
+boolean = _Boolean()
```
Comment (fehiepsi) [inline numpyro/distributions/constraints.py]: Nit: we can make this _IntegerInterval(0, 1)

File: numpyro/distributions/discrete.py (hunk 1 above)
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: That would be one way to do it, but there are many cases where we can only do validation after initialization, e.g. for binomial, where support is in `[0, n]`. Also the multivariate cases are a mess to handle, so I think it would be preferable prioritizing #82 than invest more time in working around these issues. 

For the new distribution classes, I currently have `probs` and `logits` as separate parametrizations for the same reason, and it is worth thinking if it is better to separate out the implementations.

Comment (fehiepsi) [reply numpyro/distributions/discrete.py]: About support, I think we just mimic the way we did in continuous distribution: get arg and verify. I think it is what PyTorch does (except @lazy decorator). For special case like binomial, we can just overwrite that `_support` property I guess.

File: numpyro/distributions/constraints.py (hunk 3 above)
Comment (neerajprad) [reply numpyro/distributions/constraints.py]: The issue is that it will be much slower, almost 5X or so.

File: numpyro/distributions/discrete.py (hunk 1 above)
Comment (neerajprad) [reply numpyro/distributions/discrete.py]: `_support` needs to be defined at the correct place (not in gen, but rather in frozen) because it needs to be treated as a property and not something into which arguments will be passed into. In any case, this isn't an issue with the classes in `contrib`, so I'm just focusing on that for the moment.

Comment (fehiepsi) [reply numpyro/distributions/discrete.py]: Ah, I mean to modify the _support method in gen. Anyway, I’ll do that instead. :)
